### PR TITLE
DAOS-11137 cart: Set UCX env var to allow memory allocation.

### PR DESCRIPTION
**Pull request by jgmoore-or**
This PR allows DAOS with UCX to work with huge pages (allocated by the engine when NVMe storage is used).

Signed-off-by: Joseph Moore <joseph.moore@intel.com>

### CODE REVIEW

File: src/cart/crt_init.c [hunk 1]
```diff
@@ -116,6 +118,7 @@ static int data_init(int server, crt_init_options_t *opt)
 
 	D_DEBUG(DB_ALL, "initializing crt_gdata...\n");
 
+	setenv("UCX_IB_FORK_INIT", "n", 1);
```
Comment (johannlombardi) [inline src/cart/crt_init.c:121]: Should be set only for servers IMO.

**Review by johannlombardi — CHANGES_REQUESTED** [review]

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

File: src/cart/crt_init.c [hunk 2]
```diff
@@ -116,6 +118,9 @@ static int data_init(int server, crt_init_options_t *opt)
 
 	D_DEBUG(DB_ALL, "initializing crt_gdata...\n");
 
+	if (crt_is_service())
+		setenv("UCX_IB_FORK_INIT", "n", 1);
```
Comment (soumagne) [inline src/cart/crt_init.c:122]: I think you would also want to check whether or not UCX is being used. It would just be cleaner in my opinion to keep OFI and UCX env settings separate. Also testing before if it has already been set and printing a message would also be helpful I think before you override it.

File: src/cart/crt_init.c [hunk 3]
```diff
@@ -29,7 +29,9 @@ dump_envariables(void)
 		"OFI_PORT", "OFI_INTERFACE", "OFI_DOMAIN", "CRT_CREDIT_EP_CTX",
 		"CRT_CTX_SHARE_ADDR", "CRT_CTX_NUM", "D_FI_CONFIG",
 		"FI_UNIVERSE_SIZE", "CRT_ENABLE_MEM_PIN",
-		"FI_OFI_RXM_USE_SRX", "D_LOG_FLUSH", "CRT_MRC_ENABLE" };
+		"FI_OFI_RXM_USE_SRX", "D_LOG_FLUSH", "CRT_MRC_ENABLE",
+		"RDMAV_HUGEPAGES_SAFE", "UCX_LOG_LEVEL", "UCX_MEM_LOG_LEVEL",
+		"UCX_IB_FORK_INIT" };
```
Comment (soumagne) [inline src/cart/crt_init.c:34]: Same here as other comment, I would separate UCX from OFI env variables that we dump instead of having that huge list of env vars. If you think that should be done in a separate PR, it would be good to enter an issue for it.

**Comment by johannlombardi** [comment]
Please update the patch as per review feedback.

**Comment by jgmoore-or** [comment]
Just now discovered that crt_is_service check is using state that is not yet initialized where this env var setting is being done.  This will have to be fixed to make the setting happen on server side.

**Comment by jgmoore-or** [comment]
Some minor issues with the dump env variable function are currently present (FI_UNIVERSE_SIZE is printed, but then changed later in the function, for example).

I think this PR should do the minimum, and a subsequent patch can be done to clean up the overall environment variable management.  I'll leave the final decision up to the reviewers.

File: src/cart/crt_init.c (hunk 2 above)
Comment (jgmoore-or) [reply src/cart/crt_init.c]: I'll need to check if the provider has been established at this point in the code, if we need to not set this on servers running OFI (it shouldn't cause a problem to set it when OFI is used).  I pulled the UCX env vars out of the dump function for now, thinking that your suggestion of a subsequent env var clean up patch makes better sense.

**Comment by jgmoore-or** [comment]
I've got some work left to properly to do the get and print request and keep with the approach used for other variables.

Hence I've reconverted to draft.

**Comment by jgmoore-or** [comment]
No longer a draft.

**Review by daltonbohning — COMMENTED** [review]
I'm not familiar enough with this code to know if it's correct or not

**Review by soumagne — APPROVED** [review]

File: src/cart/crt_init.c (hunk 3 above)
Comment (jgmoore-or) [reply src/cart/crt_init.c]: I removed all UCX environment variables from the list for this PR.  I agree with the suggestion to do the refactor in a subsequent PR.

**Review by johannlombardi — APPROVED** [review]
